go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy_controller.go

```diff
@@ -0,0 +1,509 @@
+		"baselineAdminNetworkPolicy",
+	)
+	_, err = banpInformer.Informer().AddEventHandler(factory.WithUpdateHandlingForObjReplace(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onBANPAdd,
+		UpdateFunc: c.onBANPUpdate,
+		DeleteFunc: c.onBANPDelete,
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("could not add Event Handler for banpInformer during admin network policy controller initialization, %w", err)
+	}
+
+	klog.Info("Setting up event handlers for Namespaces in Admin Network Policy controller")
+	c.anpNamespaceLister = namespaceInformer.Lister()
+	c.anpNamespaceSynced = namespaceInformer.Informer().HasSynced
+	c.anpNamespaceQueue = workqueue.NewNamedRateLimitingQueue(
+		workqueue.NewItemFastSlowRateLimiter(1*time.Second, 5*time.Second, 5),
+		"anpNamespaces",
+	)
+	_, err = namespaceInformer.Informer().AddEventHandler(factory.WithUpdateHandlingForObjReplace(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onANPNamespaceAdd,
+		UpdateFunc: c.onANPNamespaceUpdate,
+		DeleteFunc: c.onANPNamespaceDelete,
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("could not add Event Handler for namespace Informer during admin network policy controller initialization, %w", err)
+	}
+
+	klog.Info("Setting up event handlers for Pods in Admin Network Policy controller")
+	c.anpPodLister = podInformer.Lister()
+	c.anpPodSynced = podInformer.Informer().HasSynced
+	c.anpPodQueue = workqueue.NewNamedRateLimitingQueue(
+		workqueue.NewItemFastSlowRateLimiter(1*time.Second, 5*time.Second, 5),
+		"anpPods",
+	)
+	_, err = podInformer.Informer().AddEventHandler(factory.WithUpdateHandlingForObjReplace(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onANPPodAdd,
+		UpdateFunc: c.onANPPodUpdate,
+		DeleteFunc: c.onANPPodDelete,
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("could not add Event Handler for pod Informer during admin network policy controller initialization, %w", err)
+	}
+
+	// TODO(tssurya): We don't use recorder now but will add events in future iterations
+	c.eventRecorder = recorder
+
+	return c, nil
+}
+
+// Run will not return until stopCh is closed. workers determines how many
+// objects (pods, namespaces, anps, banps) will be handled in parallel.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting controller %s", c.controllerName)
+
+	// Wait for the caches to be synced
+	klog.Info("Waiting for informer caches to sync")
+	if !util.WaitForNamedCacheSyncWithTimeout(c.controllerName, stopCh, c.anpCacheSynced, c.banpCacheSynced, c.anpNamespaceSynced, c.anpPodSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		klog.Errorf("Error syncing caches for admin network policy and baseline admin network policy")
+		return
+	}
+
+	klog.Infof("Repairing Admin Network Policies")
+	// Run the repair function at startup so that we synchronize KAPI and OVNDBs
+	err := c.repairAdminNetworkPolicies()
```

Comment:
there may be a problem with how we use repair, because policy changes may happen between `c.anpLister.List(labels.Everything())` call in the `repairAdminNetworkPolicies` and `Get` calls from the `syncAdminNetworkPolicy`. 

So if ANP is deleted between these 2 points, it won't be cleaned up by the the repair, because it existed at that point, but when `syncAdminNetworkPolicy` is called it already may be deleted. Then `clearAdminNetworkPolicy` will check that ANP is not in cache and return. We can fix that by removing that early return from `clearAdminNetworkPolicy` and make a full cleanup only based on ANP name. Then the next problem is address sets cleanup: if policy is updated between these 2 points (instead of deleted), it won't cleanup address sets in the repair, and it won't do that in `createNewANP` either, and that may be a bit more difficult to fix.

We can also just say that this case when updates happen during (re)start are rare and ignore this :)

Reply:
> there may be a problem with how we use repair, because policy changes may happen between `c.anpLister.List(labels.Everything())` call in the `repairAdminNetworkPolicies` and `Get` calls from the `syncAdminNetworkPolicy`.
> 
> So if ANP is deleted between these 2 points, it won't be cleaned up by the the repair, because it existed at that point, but when `syncAdminNetworkPolicy` is called it already may be deleted. Then `clearAdminNetworkPolicy` will check that ANP is not in cache and return. We can fix that by removing that early return from `clearAdminNetworkPolicy` and make a full cleanup only based on ANP name. Then the next problem is address sets cleanup: if policy is updated between these 2 points (instead of deleted), it won't cleanup address sets in the repair, and it won't do that in `createNewANP` either, and that may be a bit more difficult to fix.
> 
> We can also just say that this case when updates happen during (re)start are rare and ignore this :)

nice catch nadia! the only reason why i had added that check is for anp's with duplicate prio's or prio with 99 whose setup is not done, then nothing to do, same with a quick create/delete where by the setup was never done since lock was held by a pod update or something - also for symmetry reasons, feels bad to use cache for create/update and not use it for delete... let me leave a comment pointing to this discussion so that these two scenarios are thought of.

but i think the timing between those List and Get calls should be really negligible for anyone to be able to squeeze an ANP delete and update plus I really hope a cluster admin doesn't choose the ovnkube upgrade/reboot/restart events to really delete and update objects...

Reply:
This would work better if repair didn't care about existing policies at all and that was handled in `syncAdminNetworkPolicy`
https://github.com/ovn-org/ovn-kubernetes/pull/3659#discussion_r1285695187

Reply:
ack going ahead with dealing with only deletion of AS-es and PGs for non existing policies in repair
and then moving all the per rule change calls into createNewANP so that both the AS/ACL changes are reflected for each new addition.